arch/xtensa/src/esp32/esp32_ble_adapter.c

```diff
@@ -0,0 +1,2817 @@
+    {
+      wlerr("Invalid controller status");
+      return -1;
+    }
+
+  if (btdm_osi_funcs_register(&g_osi_funcs) != 0)
+    {
+      wlinfo("Error, probably invalid OSI Functions\n");
+      return -EINVAL;
+    }
+
+  wlinfo("BT controller compile version [%s]\n",
+                              btdm_controller_get_compile_version());
+
+  /* If all the bt available memory was already released,
+   * cannot initialize bluetooth controller
+   */
+
+  if (btdm_dram_available_region[0].mode == ESP_BT_MODE_IDLE)
+    {
+      wlerr("Error, bt available memory was released\n");
+      return ESP_ERR_INVALID_STATE;
+    }
+
+  if (cfg == NULL)
+    {
+      wlerr("%s  %d\n", __func__, __LINE__);
+      return -EINVAL;
+    }
+
+  cfg->controller_task_stack_size  = CONFIG_ESP32_BLE_TASK_STACK_SIZE;
+  cfg->controller_task_prio        = CONFIG_ESP32_BLE_TASK_PRIORITY;
+
+  cfg->bt_max_sync_conn            = CONFIG_ESP32_BLE_MAX_CONN;
+  cfg->magic                       = ESP_BT_CONTROLLER_CONFIG_MAGIC_VAL;
+
+  if (((cfg->mode & ESP_BT_MODE_BLE) && (cfg->ble_max_conn <= 0 ||
+        cfg->ble_max_conn > BTDM_CONTROLLER_BLE_MAX_CONN_LIMIT)) ||
+      ((cfg->mode & ESP_BT_MODE_CLASSIC_BT) && (cfg->bt_max_acl_conn <= 0 ||
+        cfg->bt_max_acl_conn > BTDM_CONTROLLER_BR_EDR_MAX_ACL_CONN_LIMIT)) ||
+      ((cfg->mode & ESP_BT_MODE_CLASSIC_BT) &&
+       (cfg->bt_max_sync_conn > BTDM_CONTROLLER_BR_EDR_MAX_SYNC_CONN_LIMIT)))
+    {
+      wlerr("%s  %d\n", __func__, __LINE__);
+      return -EINVAL;
+    }
+
+  wlinfo("BT controller compile version [%s]",
+         btdm_controller_get_compile_version());
+
+  btdm_controller_mem_init();
+
+  wlinfo("Memory initialized!\n");
+
+  bt_periph_module_enable(PERIPH_BT_MODULE);
+
+  /* set default sleep clock cycle and its fractional bits */
+
+  g_btdm_lpcycle_us_frac = RTC_CLK_CAL_FRACT;
+  g_btdm_lpcycle_us = 2 << (g_btdm_lpcycle_us_frac);
+
+  btdm_controller_set_sleep_mode(BTDM_MODEM_SLEEP_MODE_NONE);
+
+  btdm_cfg_mask = btdm_config_mask_load();
+
+  wlinfo("Going to call btdm_controller_init\n");
+
+  if (btdm_controller_init(btdm_cfg_mask, cfg) != 0)
+    {
+      err = -ENOMEM;
+      goto error;
+    }
+
+  wlinfo("The btdm_controller_init was initialized\n");
+
+#ifdef CONFIG_BTDM_COEX_BLE_ADV_HIGH_PRIORITY
+  coex_ble_adv_priority_high_set(true);
+#else
+  coex_ble_adv_priority_high_set(false);
+#endif
+
+  btdm_controller_status = ESP_BT_CONTROLLER_STATUS_INITED;
+
+  return OK;
+
+error:
+  return err;
+}
+
+/****************************************************************************
+ * Name: esp32_bt_controller_deinit
+ *
+ * Description:
+ *    Deinit  BT controller.
+ * Input Parameters:
+ *    None
+ *
+ * Returned Value:
+ *    None
+ *
+ ****************************************************************************/
+
+int esp32_bt_controller_deinit(void)
+{
+  if (btdm_controller_status != ESP_BT_CONTROLLER_STATUS_INITED)
+    {
+      return -1;
+    }
+
+  btdm_controller_deinit();
```

Comment:
Is it necessary to deinit? The check before is not intended to check if it was previously initialized?